File: python/parser.py
import re
from typing import List
from .domain import FailureContext

class Parser:
    @staticmethod
    def parse(raw_log: str) -> FailureContext:
        failed_tests: List[str] = []
        stack_trace: List[str] = []
        failure_location = None
        failure_reason = None

        lines = raw_log.split('\n')
        in_failures = False

        for line in lines:
            if "=== FAILURES ===" in line:
                in_failures = True
                continue
            if "=== short test summary info ===" in line:
                in_failures = False
                continue

            if in_failures:
                # Capture test function headers
                match_fn = re.match(r'^_{3,}\s+(\w+)\s+_{3,}$', line.strip())
                if match_fn:
                    failed_tests.append(match_fn.group(1))

                # Capture stack trace locations like file.py:line
                match_loc = re.match(r'^([\w\/\.\-]+:\d+):?\s*(.*)$', line.strip())
                if match_loc and not failure_location:
                    failure_location = match_loc.group(1)

                # Capture Error lines like E   ValueError: ...
                if line.strip().startswith("E "):
                    failure_reason = line.strip()[2:].strip()

                stack_trace.append(line)

            # Fallback capture from short test summary
            if line.startswith("FAILED "):
                parts = line.split(" - ")
                if len(parts) > 0 and not failed_tests:
                    failed_tests.append(parts[0].replace("FAILED ", "").strip())
                if len(parts) > 1 and not failure_reason:
                    failure_reason = parts[1].strip()

        return FailureContext(
            raw_log=raw_log,
            failed_tests=failed_tests,
            failure_location=failure_location or "unknown:0",
            failure_reason=failure_reason or "Unknown test failure",
            stack_trace=stack_trace
        )

File: src/parser.ts
import { FailureContext } from './domain';

export class Parser {
  static parse(rawLog: string): FailureContext {
    const context: FailureContext = {
      rawLog,
      sourceLocations: [],
      relevantLogLines: []
    };

    if (!rawLog || typeof rawLog !== 'string') {
      return context;
    }

    const lines = rawLog.split('\n');

    // 1. Extract test name
    const testNameMatch = rawLog.match(/_{3,} (test_[a-zA-Z0-9_]+) _{3,}/);
    if (testNameMatch) {
      context.testName = testNameMatch[1];
    } else {
      const failedMatch = rawLog.match(/FAILED\s+[^:]+::(test_[a-zA-Z0-9_]+)/);
      if (failedMatch) {
        context.testName = failedMatch[1];
      }
    }

    // 2. Extract Exception type and message
    // Usually pytest prefixes lines with "E   "
    for (const line of lines) {
      const errorMatch = line.match(/^E\s+([A-Za-z0-9_]+(?:Error|Exception)):\s*(.*)/);
      if (errorMatch) {
        if (!context.errorType) {
          context.errorType = errorMatch[1];
          context.errorMessage = errorMatch[2].trim();
        }
        context.relevantLogLines.push(line);
      } else if (line.match(/^E\s+/)) {
        context.relevantLogLines.push(line);
      } else if (line.match(/^>\s+/)) {
        context.relevantLogLines.push(line);
      }
    }

    // Fallback if not found with 'E   '
    if (!context.errorType) {
      for (let i = lines.length - 1; i >= 0; i--) {
        const line = lines[i];
        const errorMatch = line.match(/^([A-Za-z0-9_]+(?:Error|Exception)):\s*(.*)/);
        if (errorMatch && !line.startsWith('E ')) {
          context.errorType = errorMatch[1];
          context.errorMessage = errorMatch[2].trim();
          break;
        }
      }
    }

    // 3. Extract source locations
    const sourceLocations = new Set<string>();
    for (const line of lines) {
      const locationMatch = line.match(/([a-zA-Z0-9_\-\./\\]+\.py):(\d+):/);
      if (locationMatch) {
        sourceLocations.add(`${locationMatch[1]}:${locationMatch[2]}`);
      } else {
        const tbMatch = line.match(/File "([^"]+)", line (\d+)/);
        if (tbMatch) {
          sourceLocations.add(`${tbMatch[1]}:${tbMatch[2]}`);
        }
      }
    }
    context.sourceLocations = Array.from(sourceLocations);

    // 4. Extract stack trace block (rudimentary)
    const traceStartIndex = lines.findIndex(l => 
      l.includes('Traceback (most recent call last):') || 
      l.match(/_{3,} test_/) ||
      l.startsWith('FAILED ')
    );
    
    if (traceStartIndex !== -1) {
      const traceLines = [];
      for (let i = traceStartIndex; i < lines.length; i++) {
        const l = lines[i];
        if (l.startsWith('===') || l.match(/^-{2,} short test summary info/)) {
          break;
        }
        traceLines.push(l);
      }
      if (traceLines.length > 0) {
        context.stackTrace = traceLines.join('\n');
      }
    }

    return context;
  }
}
